eval/bundled/epm.elv.go

```diff
@@ -1,111 +1,279 @@
+      } else {
+        fail "No handler for '"$what"' defined in method '"$method"'"
+      }
+    } else {
+      fail "No handler defined for method '"$method"', specified in in config file "(-domain-config-file $dom)
     }
-    dest = $-data-dir/lib/$pkg
-    if (not ?(test -d $dest)) {
-        -error 'Package '$pkg' does not exist locally.'
-        return
+  }
+}
+
+# Read and parse the package metadata, if it exists
+fn metadata [pkg]{
+  res = [&]
+  mdata = (-package-metadata-file $pkg)
+  if (and (is-installed $pkg) ?(test -f $mdata)) {
+    res = (cat $mdata | from-json)
+  }
+  put $res
+}
+
+# Print out information about a package
+fn query [pkg]{
+  data = (metadata $pkg)
+  echo (edit:styled "Package "$pkg cyan)
+  echo (edit:styled "Source:" blue) (-package-method $pkg) (-package-op $pkg src)
+  if (is-installed $pkg) {
+    echo (edit:styled "Installed at "(dest $pkg) green)
+  } else {
+    echo (edit:styled "Not installed" red)
+  }
+  keys $data | each [key]{
+    echo (edit:styled $key":" blue) $data[$key]
+  }
+}
+
+# Uninstall a single package by removing its directory
+fn -uninstall-package [pkg]{
+  if (not (is-installed $pkg)) {
+    -error "Package "$pkg" is not installed."
+    return
+  }
+  dest = (dest $pkg)
+  -info "Removing package "$pkg
+  rm -rf $dest
+}
+
+######################################################################
+# Main user-facing functions
+
+# List installed packages
+fn installed {
+  e:ls $-lib-dir | each [dom]{
+    if ?(test -f (-domain-config-file $dom)) {
+      -read-domain-config $dom
+      lvl = $-domain-config[$dom][levels]
+      find $-lib-dir/$dom -type d -depth $lvl | each [pkg]{
+        replaces $-lib-dir/ "" $pkg
+      }
     }
-    -info 'Removing package '$pkg
-    rm -rf $dest
-    # issue #486
-    {
-        for installed $installed-pkgs {
-            if (not-eq $installed $pkg) {
-                echo $installed
-            }
-        }
-    } > $-installed
+  }
 }
 
-fn uninstall [@pkgs]{
-    if (eq $pkgs []) {
-        -error 'Must specify at least one package.'
-        return
+# Install and upgrade are method-specific, so we call the
+# corresponding functions using -package-op
+fn install [&silent-if-installed=$false @pkgs]{
+  for pkg $pkgs {
+    if (is-installed $pkg) {
+      if (not $silent-if-installed) {
+        -info "Package "$pkg" is already installed."
+      }
+    } else {
+      -package-op $pkg install
     }
-    for pkg $pkgs {
-        -uninstall-one $pkg
+  }
+}
+
+fn upgrade [@pkgs]{
+  if (eq $pkgs []) {
+    pkgs = [(installed)]
+    -info 'Upgrading all installed packages'
+  }
+  for pkg $pkgs {
+    if (not (is-installed $pkg)) {
+      -error "Package "$pkg" is not installed."
+    } else {
+      -package-op $pkg upgrade
     }
+  }
 }
-`
+
+# Uninstall is the same for everyone, just remove the directory
+fn uninstall [@pkgs]{
+  if (eq $pkgs []) {
+    fail 'Must specify at least one package.'
```

Comment:
Use "-error" to be consistent with other commands.

Reply:
Makes sense. `-error` for "regular" errors (usage) and `fail` for "internal" errors (lack of default config for a domain, etc.)